### PR TITLE
Replace CombinedPublicKey reference in MeasurementSpec with Duchy entries in Requisition.

### LINKED ISSUE #4: Support use of new ElGamal keys for each Measurement
The current API design isn't well-suited to allow a new set of ElGamal keys for each Measurement. The top-level `CombinedPublicKeys` resource implies reuse of these keys across Measurements instead. This is not a good cryptographic practice.

### PR DESCRIPTION
**Pull request by SanjayVas**
This adds the Duchy resource type and removes the CombinedPublicKey resource type.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)

### CODE REVIEW

File: src/main/proto/wfa/measurement/api/v2alpha/duchy.proto [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  // X.509 certificate in DER format which can be verified using the `Duchy`'s
+  // root certificate. Required.
+  bytes certificate = 2;
+
+  // Resource key of the `DuchyCertificate` that matches `certificate`.
+  // Output-only.
+  DuchyCertificate.Key duchy_certificate = 3;
+}
+
+message DuchyCertificate {
```
Comment (efoxepstein) [inline src/main/proto/wfa/measurement/api/v2alpha/duchy.proto:40]: What is the purpose of this resource? It seems to hold the same data as `Duchy`. What's its relationship to Duchy? 1-to-1? Many-to-1? Just a view of Duchy with a different name?

File: src/main/proto/wfa/measurement/api/v2alpha/requisition.proto [hunk 2]
```diff
@@ -58,6 +59,22 @@ message Requisition {
   // `DataProviderEntry` in `measurement`.  Required. Immutable.
   bytes encrypted_requisition_spec = 6;
 
+  message DuchyEntry {
+    // Resource key of the `Duchy` that this entry is for. Required. Immutable.
+    Duchy.Key duchy = 1;
+
+    // Resource key of the `DuchyCertificate` for this entry. Required.
+    // Immutable.
+    DuchyCertificate.Key duchy_certificate = 2;
```
Comment (efoxepstein) [inline src/main/proto/wfa/measurement/api/v2alpha/requisition.proto:68]: Can't this be inferred from Duchy.Key? (Duchy.duchy_certificate field of the Duchy)

**Review by efoxepstein — CHANGES_REQUESTED** [review]

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @efoxepstein and @kungfucraig)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 40 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-40:-MX8o7tr5cCCJ23N2mmb:b7irffq) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L40)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

What is the purpose of this resource? It seems to hold the same data as `Duchy`. What's its relationship to Duchy? 1-to-1? Many-to-1? Just a view of Duchy with a different name?
</blockquote></details>

The relationship is one `Duchy` to one or more `DuchyCertificate`s, where the `Duchy` contains a view to one of them. This structure guarantees that a certificate always exists for the resource.

This is the same pattern as `DataProviderCertificate` and `MeasurementConsumerCertificate`.

We can be more specific about which certificate the parent resource has a view of to provide more usefulness/determinism. E.g.  the certificate with the longest TTL.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 68 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHq-r1-68:-MX8pMnOA_tqw-nfUZQ9:b3bmtny) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L68)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Can't this be inferred from Duchy.Key? (Duchy.duchy_certificate field of the Duchy)
</blockquote></details>

No, as this can be any `DuchyCertificate` and not just the one that the `Duchy` resource has a view of.

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 0 of 9 files reviewed, 4 unresolved discussions (waiting on @kungfucraig and @SanjayVas)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 33 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8s8De5OxVMYW5LNcO:-MX8s8De5OxVMYW5LNcP:b-3h4lfw) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L33)):*
> ```protobuf
>   // X.509 certificate in DER format which can be verified using the `Duchy`'s
>   // root certificate. Required.
>   bytes certificate = 2;
> ```

How does this field work? Why isn't it output-only? Is this how Duchies update their certs? (What does it mean to change this field but not duchy_certificate?)
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 40 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-40:-MX8rjawCaYtCvOfjgoY:bcm4x79) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L40)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

The relationship is one `Duchy` to one or more `DuchyCertificate`s, where the `Duchy` contains a view to one of them. This structure guarantees that a certificate always exists for the resource.

This is the same pattern as `DataProviderCertificate` and `MeasurementConsumerCertificate`.

We can be more specific about which certificate the parent resource has a view of to provide more usefulness/determinism. E.g.  the certificate with the longest TTL.
</blockquote></details>

Documenting this would be helpful for me. What about just naming the fields `latest_certificate` or `active_certificate` or something like that, to indicate that it is a "blessed" one of many?
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 49 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8sQwbBxZc8FWPM5f0:-MX8sQwbBxZc8FWPM5f1:b4q1keg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L49)):*
> ```protobuf
>   // Resource key.
>   Key key = 1;
> ```

For debugging purposes, would it make sense to embed a creation timestamp in here? Or some human-readable "name"? (Or is that embedded in the DER certificate? If so, is it worth denormalizing that for debuggability into the proto itself?)
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 68 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHq-r1-68:-MX8rXLfC6TjXGg_fnwk:bgfuf4o) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L68)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

No, as this can be any `DuchyCertificate` and not just the one that the `Duchy` resource has a view of.
</blockquote></details>

In that case, can't the Duchy.Key be inferred from the DuchyCertificate.Key? Or is this to ensure that the Keys are treated as opaque by clients?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 0 of 9 files reviewed, 4 unresolved discussions (waiting on @efoxepstein and @kungfucraig)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 33 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8s8De5OxVMYW5LNcO:-MX9431J4VAZYYZvbX_E:b2k12sp) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L33)):*
For resources that are creatable via the API, this is how one specifies the initial certificate for creation. For resources that are updatable via the API, changing this field could result in the creation of a new *Certificate resource.

> What does it mean to change this field but not duchy_certificate?

It's not possible. `duchy_certificate` is output-only, and the field will always be set to the matching resource key for any `Duchy` resource returned by the API.
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 40 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-40:-MX96Sj58fml3hRtmF42:bp21cjf) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L40)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Documenting this would be helpful for me. What about just naming the fields `latest_certificate` or `active_certificate` or something like that, to indicate that it is a "blessed" one of many?
</blockquote></details>

Well, right now it's just *any* `DuchyCertificate`. I can make the semantic and naming change in a subsequent PR for everything that follows this pattern at once.
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 49 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8sQwbBxZc8FWPM5f0:-MX94ma8EVoyf5LTkAUo:bgojg3h) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L49)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

For debugging purposes, would it make sense to embed a creation timestamp in here? Or some human-readable "name"? (Or is that embedded in the DER certificate? If so, is it worth denormalizing that for debuggability into the proto itself?)
</blockquote></details>

The timetamps for the valid range are embedded in the X.509 certificate. I've thought about adding an X509Certificate message with various output-only fields denormalized from the DER bytes, but we haven't needed it yet. I personally don't think it's worth adding just for debugging.
___
*[src/main/proto/wfa/measurement/api/v2alpha/requisition.proto, line 68 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHq-r1-68:-MX97-uyBIwi7REFenIw:b-b6qp33) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/requisition.proto#L68)):*
> Or is this to ensure that the Keys are treated as opaque by clients?

Only `Key` messages ever directly reference IDs. Everywhere else, we refer to those `Key` messages. We would do the same if using resource names as described in the AIPS.

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 0 of 9 files reviewed, all discussions resolved (waiting on @kungfucraig)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 40 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-40:-MX9CE7u5ZsreLMhnC6h:b-yxpaxo) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L40)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Well, right now it's just *any* `DuchyCertificate`. I can make the semantic and naming change in a subsequent PR for everything that follows this pattern at once.
</blockquote></details>

SGTM, can you add a TODO or make a bug?
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 49 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8sQwbBxZc8FWPM5f0:-MX9CH4--Pm6xLXXLFkv:bmcx69o) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L49)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

The timetamps for the valid range are embedded in the X.509 certificate. I've thought about adding an X509Certificate message with various output-only fields denormalized from the DER bytes, but we haven't needed it yet. I personally don't think it's worth adding just for debugging.
</blockquote></details>

Creation time is different from the timestamps in the certificate.

We don't necessarily need to expose it in the public API, but it could be very useful in debugging. Adding a malicious certificate is certainly an attack vector (how effective it would be is up in the air), so having a little bit more info about it could help. For example, one could imagine that we need to rate-limit certificate creation or notify Duchies when their certificates haven't been updated in a while or something.  Both of these would be enabled by having a create time.

Leaving it out of the public API for now SGTM

File: src/main/proto/wfa/measurement/api/v2alpha/duchy.proto [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+syntax = "proto3";
+
+package wfa.measurement.api.v2alpha;
+
+option java_package = "org.wfanet.measurement.api.v2alpha";
+option java_multiple_files = true;
+option java_outer_classname = "DuchyProto";
+
+// Resource message representing a Duchy.
+message Duchy {
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/api/v2alpha/duchy.proto:24]: What is the purpose of this?

**Review by kungfucraig — CHANGES_REQUESTED** [review]

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 0 of 9 files reviewed, 1 unresolved discussion (waiting on @kungfucraig)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 24 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-24:-MXDJ3FD-wv65F-WHyZp:b-5b0jc3) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L24)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

What is the purpose of this?
</blockquote></details>

As indicated in the subject, this PR removes the `CombinedPublicKey` reference in `MeasurementSpec` and replaces it with a map of `Duchy` to signed ElGamal public key in `Requisition`.
___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 40 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-40:-MXDKB6q-Tu3oSUcemeW:b-bzuzog) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L40)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

SGTM, can you add a TODO or make a bug?
</blockquote></details>

Already drafting the next PR.

**Review by efoxepstein — DISMISSED** [review]
> Reviewed 7 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 7 of 9 files reviewed, 1 unresolved discussion (waiting on @kungfucraig)

**Review by SanjayVas — COMMENTED** [review]
> Dismissed @kungfucraig from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 7 of 9 files reviewed, all discussions resolved (waiting on @kungfucraig)

File: src/main/proto/wfa/measurement/api/v2alpha/duchy.proto (hunk 1 above)
Comment (zachcwc) [reply src/main/proto/wfa/measurement/api/v2alpha/duchy.proto:40]: I believe Sanjay answered here https://github.com/world-federation-of-advertisers/cross-media-measurement-api/pull/5#pullrequestreview-625666490

Comment (SanjayVas) [reply src/main/proto/wfa/measurement/api/v2alpha/duchy.proto:40]: Thanks Zach. See the thread replies from Reviewable.io and ignore these.

**Review by zachcwc — DISMISSED** [review]

**Review by SanjayVas — COMMENTED** [review]
> Dismissed @kungfucraig from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5)* status: 7 of 9 files reviewed, all discussions resolved (waiting on @SanjayVas)

___
*[src/main/proto/wfa/measurement/api/v2alpha/duchy.proto, line 53 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement-api/5#-MX8o6IQymjZ66v-MbHo-r1-53:-MXicaUD9JfNbyBR9rib:b-e37bjm) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/5c2e069ce08097f46ebd4e0dc7de39333507e595/src/main/proto/wfa/measurement/api/v2alpha/duchy.proto#L53)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.

Note that due to our formatters/linters not caring about this, it should be considered a "no-comment" issue. For example, see go/java-style#s8.1.1-no-comment-issues (Google-internal).

**Review by kungfucraig — APPROVED** [review]

**Pull request merged**